src/stores/TypingStore.js

```diff
@@ -0,0 +1,94 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import MatrixClientPeg from "../MatrixClientPeg";
+import SettingsStore from "../settings/SettingsStore";
+import Timer from "../utils/Timer";
+
+const TYPING_USER_TIMEOUT = 10000;
+const TYPING_SERVER_TIMEOUT = 30000;
+
+/**
+ * Tracks typing state for users.
+ */
+export default class TypingStore {
+    constructor() {
+        this.reset();
+    }
+
+    static sharedInstance(): TypingStore {
+        if (global.mxTypingStore === undefined) {
+            global.mxTypingStore = new TypingStore();
+        }
+        return global.mxTypingStore;
+    }
+
+    /**
+     * Clears all cached typing states. Intended to be called when the
+     * MatrixClientPeg client changes.
+     */
+    reset() {
+        this._typingStates = {
+            // "roomId": {
+            //     isTyping: bool,     // Whether the user is typing or not
+            //     userTimer: Timer,   // Local timeout for "user has stopped typing"
+            //     serverTimer: Timer, // Maximum timeout for the typing state
+            // },
+        };
+    }
+
+    /**
+     * Changes the typing status for the MatrixClientPeg user.
+     * @param {string} roomId The room ID to set the typing state in.
+     * @param {boolean} isTyping Whether the user is typing or not.
+     */
+    setSelfTyping(roomId: string, isTyping: boolean): void {
+        if (!SettingsStore.getValue('sendTypingNotifications')) return;
+        if (SettingsStore.getValue('lowBandwidth')) return;
+
+        let currentTyping = this._typingStates[roomId];
+        if ((!isTyping && !currentTyping) || (currentTyping && currentTyping.isTyping === isTyping)) {
+            // No change in state, so don't do anything. We'll let the timer run its course.
+            return;
+        }
+
+        if (!currentTyping) {
+            currentTyping = this._typingStates[roomId] = {
+                isTyping: isTyping,
+                serverTimer: new Timer(TYPING_SERVER_TIMEOUT),
+                userTimer: new Timer(TYPING_USER_TIMEOUT),
+            };
+        }
+
+        currentTyping.isTyping = isTyping;
+
+        if (isTyping) {
+            currentTyping.serverTimer.restart().finished().then(() => {
+                const currentTyping = this._typingStates[roomId];
+                if (currentTyping) currentTyping.isTyping = false;
+
+                // The server will (should) time us out on typing, so we don't
+                // need to advertise a stop of typing.
+            });
+
+            currentTyping.userTimer.restart().finished().then(() => {
```

Comment:
Better check if the timer is not already running (and hence has a resolve handler attached already), so that the code below won't run multiple times when the timer expires. E.g. by putting `finished().then(...` before `restart()` and wrap it in `if (!timer.isRunning())`.